File: task3.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>
#include <locale.h>
using namespace std;

void fillingTheVector(vector<int>& vec) //заполнение вектора случайными числами
{ 
    random_device rd;//random_device, который является источником недетерминированных случайных чисел.
    //Затем мы используем это устройство для заполнения генератора std::minstd_rand. Затем функция генератора() используется для генерации случайного числа
    minstd_rand generator(rd());
    uniform_int_distribution<int> distribution(0, 100);// функция destribition для задания диапозона значений

    for (int& item : vec) 
    {
        item = distribution(generator);
    }
}

int ElementCountInVector(const vector<int>& container, int element) //подсчет количества вхождений заданного элемента в вектор
{ 
    int count = 0;
    for (const int& value : container) 
    {
        if (value == element) count++;
    }
    return count;
}

double mathExpectation(vector<int> container) //вычисление реального математического ожидания
{ 
    float Mx = 0.0;
    for (auto i = 0; i < container.size(); i++)
    {
        Mx += container[i] * 0.01;// диапозон чисел 100. Каждое равновероятно
    }
    return Mx;
}

double mathExp(vector <int> container) //вычисление ожидаемого математического ожидания
{
    float Mx0 = 0.0;
    Mx0 = container.size()/2;//  при равномерном распределении случайной величины в диапазоне от a до b, математическое ожидание вычисляется как среднее арифметическое этих границ
    return Mx0;
}

double mathDispersion(vector<int> container) //функция для вычисления дисперсии
{
    float Mx = mathExpectation(container);// вычисляем дисперсию относительно реального мат ожидания
    float sum = 0.0;
    for (int i = 0; i < container.size(); i++)
    {
        sum += pow(container[i] - Mx, 2);// суммируем разницу между всеми значениями и мат ожиданием в квадрате
    }
    return sum / (container.size()-1);//делим на количество в выборке- 1
}

double tabLaplas(double x) //функции Лапласа (стандартного нормального распределения) для заданного аргумента x
{ 
    return (1.0 / 2.0) * (1.0 + erf(x / sqrt(2.0)));
}

double chiSquare(const vector<int>& container)
{
    double Mx = mathExpectation(container);
    double chi = 0.0;
    for (int i = 0; i < 10; i++) // Перебор 10 интервалов
    {
        int observed = ElementCountInVector(container, i * 10); // Наблюдаемая частота
        double expected = container.size() * 0.1; // Ожидаемая частота

        chi += pow(observed - expected, 2) / expected;
    }
    return chi;
}
double quantile_chi_square(double p, int df)
{
    // Функция для вычисления квантили распределения хи-квадрат
    // Реализация на основе библиотеки math.h
    return sqrt(2.0 * tgamma(df / 2.0) * pow(p, 1.0 / df) / tgamma((df - 1.0) / 2.0));
}
double chiSquareCriticalValue(int df, double alpha) 
{
    // Функция для вычисления критического значения хи-квадрат
    double p = 1.0 - alpha; // Вероятность, соответствующая уровню значимости
    return quantile_chi_square(p, df); // Вычисление критического значения
}



int main() 
{
    setlocale(LC_ALL, "Rus");

    vector<int> vector50(50);
    fillingTheVector(vector50);
   
    vector<int> vector100(100);
    fillingTheVector(vector100);

    vector<int> vector1000(1000);
    fillingTheVector(vector1000);
   
    double criticalValue = chiSquareCriticalValue(7, 0.05); // Критическое значение хи-квадрат для 99 степеней свободы и уровня значимости 0.05
   
   // для 50 элементов
    double chiSquareValue = chiSquare(vector50);
    double meanExpected = mathExpectation(vector50); //реальное математическое ожидание
    double meanObserved = mathExp(vector50);// ожидаемое мат ожидание

    cout << "X^2: " << chiSquareValue << endl;

    if (chiSquareValue <= criticalValue) 
    {
        cout << "Гипотеза о нормальном распределении не отвергается." << endl;
    }
    else 
    {
        cout << "Гипотеза о нормальном распределении отвергается." << endl;
    }

    cout << "Ожидаемое математическое ожидание: " << meanObserved << endl;
    cout << "Реальное математическое ожидание: " << meanExpected << endl;
    cout << endl;

    //для 100 элементов
    chiSquareValue = chiSquare(vector100);
    meanExpected = mathExpectation(vector100);
    meanObserved = mathExp(vector100);

    cout << "X^2: " << chiSquareValue << endl;

    if (chiSquareValue <= criticalValue) 
    {
        cout << "Гипотеза о нормальном распределении не отвергается." << endl;
    }
    else 
    {
        cout << "Гипотеза о нормальном распределении отвергается." << endl;
    }

    cout << "Ожидаемое математическое ожидание: " << meanObserved << endl;
    cout << "Реальное математическое ожидание: " << meanExpected << endl;
    cout << endl;

    //для 1000 элементов
    chiSquareValue = chiSquare(vector1000);
    meanExpected = mathExpectation(vector1000);
    meanObserved = mathExp(vector1000);

    cout << "X^2: " << chiSquareValue << endl;

    if (chiSquareValue <= criticalValue) 
    {
        cout << "Гипотеза о нормальном распределении не отвергается." << endl;
    }
    else 
    {
        cout << "Гипотеза о нормальном распределении отвергается." << endl;
    }

    cout << "Ожидаемое математическое ожидание: " << meanObserved << endl;
    cout << "Реальное математическое ожидание: " << meanExpected << endl;
    cout << endl;

    return 0;
}

File: Go/task3.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

func fillingTheVector(vec []int) { // заполнение вектора случайными числами
	rand.Seed(time.Now().UnixNano())
	for i := range vec {
		vec[i] = rand.Intn(101)
	}
}

func elementCountInVector(container []int, element int) int { // подсчет количества вхождений заданного элемента в вектор
	count := 0
	for _, value := range container {
		if value == element {
			count++
		}
	}
	return count
}

func mathExpectation(container []int) float64 { // вычисление реального математического ожидания
	var Mx float64
	for _, i := range container {
		Mx += float64(i) * 0.01 // диапазон чисел 100. Каждое равновероятно
	}
	return Mx
}

func mathExp(container []int) float64 { // вычисление ожидаемого математического ожидания
	var Mx0 float64
	Mx0 = float64(len(container)) / 2 // при равномерном распределении случайной величины в диапазоне от a до b, математическое ожидание вычисляется как среднее арифметическое этих границ
	return Mx0
}

func mathDispersion(container []int) float64 { // функция для вычисления дисперсии
	Mx := mathExpectation(container) // вычисляем дисперсию относительно реального мат ожидания
	var sum float64
	for _, i := range container {
		sum += math.Pow(float64(i)-Mx, 2) // суммируем разницу между всеми значениями и мат ожиданием в квадрате
	}
	return sum / float64(len(container)-1) // делим на количество в выборке- 1
}

func chiSquare(container []int) float64 {
	Mx := mathExpectation(container)
	var chi float64
	for i := 0; i < 10; i++ { // Перебор 10 интервалов
		observed := elementCountInVector(container, i*10) // Наблюдаемая частота
		expected := float64(len(container)) * 0.1         // Ожидаемая частота

		chi += math.Pow(float64(observed)-expected, 2) / expected
	}
	return chi
}

func main() {
	vector50 := make([]int, 50)
	fillingTheVector(vector50)

	vector100 := make([]int, 100)
	fillingTheVector(vector100)

	vector1000 := make([]int, 1000)
	fillingTheVector(vector1000)

	var criticalValue float64

	// для 50 элементов
	criticalValue = 66.338 // Критическое значение хи-квадрат для 49 степеней свободы и уровня значимости 0.05
	fmt.Println(criticalValue)
	chiSquareValue := chiSquare(vector50)
	meanExpected := mathExpectation(vector50) // реальное математическое ожидание
	meanObserved := mathExp(vector50)         // ожидаемое мат ожидание

	fmt.Printf("X^2: %f\n", chiSquareValue)

	if chiSquareValue <= criticalValue {
		fmt.Println("Гипотеза о нормальном распределении не отвергается.")
	} else {
		fmt.Println("Гипотеза о нормальном распределении отвергается.")
	}

	fmt.Printf("Ожидаемое математическое ожидание: %f\n", meanObserved)
	fmt.Printf("Реальное математическое ожидание: %f\n", meanExpected)
	fmt.Println()

	// для 100 элементов
	criticalValue = 124.342
	fmt.Println(criticalValue)
	chiSquareValue = chiSquare(vector100)
	meanExpected = mathExpectation(vector100)
	meanObserved = mathExp(vector100)

	fmt.Printf("X^2: %f\n", chiSquareValue)

	if chiSquareValue <= criticalValue {
		fmt.Println("Гипотеза о нормальном распределении не отвергается.")
	} else {
		fmt.Println("Гипотеза о нормальном распределении отвергается.")
	}

	fmt.Printf("Ожидаемое математическое ожидание: %f\n", meanObserved)
	fmt.Printf("Реальное математическое ожидание: %f\n", meanExpected)
	fmt.Println()

	// для 1000 элементов
	criticalValue = 1092.32
	fmt.Println(criticalValue)
	chiSquareValue = chiSquare(vector1000)
	meanExpected = mathExpectation(vector1000)
	meanObserved = mathExp(vector1000)

	fmt.Printf("X^2: %f\n", chiSquareValue)

	if chiSquareValue <= criticalValue {
		fmt.Println("Гипотеза о нормальном распределении не отвергается.")
	} else {
		fmt.Println("Гипотеза о нормальном распределении отвергается.")
	}

	fmt.Printf("Ожидаемое математическое ожидание: %f\n", meanObserved)
	fmt.Printf("Реальное математическое ожидание: %f\n", meanExpected)
	fmt.Println()
}
